frontend/z3_types.py

```diff
@@ -130,87 +131,61 @@ def __init__(self, config):
-                   patterns=[MultiPattern(self.subtype(x, y), self.not_subtype(y, z))], qid="not subtype"),
-
-            # a generic type is invariant
-            ForAll([x, y], Implies(self.subtype(x, self.type(y)), x == self.type(y)),
-                   patterns=[self.subtype(x, self.type(y))], qid="generic type"),
-            # a generic list type is invariant
-            ForAll([x, y], Implies(self.subtype(x, self.list(y)), x == self.list(y)),
-                   patterns=[self.subtype(x, self.list(y))], qid="generic list"),
-            # a generic set type is invariant
-            ForAll([x, y], Implies(self.subtype(x, self.set(y)), x == self.set(y)),
-                   patterns=[self.subtype(x, self.set(y))], qid="generic set"),
-            # a generic dictionary type is invariant
-            ForAll([x, y, z], Implies(self.subtype(x, self.dict(y, z)), x == self.dict(y, z)),
-                   patterns=[self.subtype(x, self.dict(y, z))], qid="generic dict")
-        ] + self.tuples_axioms() + self.functions_axioms() + self.classes_axioms(class_to_base)
+        self.subtyping = self.create_subtype_axioms(config.all_classes, type_sort)
+
+    def create_class_tree(self, all_classes, type_sort):
+        """
+        Creates a tree consisting of ClassNodes which contains all classes in all_classes,
+        where child nodes are subclasses. The root will be object.
+        """
+        tree = ClassNode('object', None, type_sort)
+        to_cover = list(all_classes.keys())
+        covered = {'object'}
+        i = 0
+        while i < len(to_cover):
+            current = to_cover[i]
+            i += 1
+            base = all_classes[current]
+            if base not in covered:
+                to_cover.append(current)
+                continue
+            base_node = tree.find(base)
+            current_node = ClassNode(current, base_node, type_sort)
+            base_node.children.append(current_node)
+            covered.add(current)
+        return tree
+
+    def create_subtype_axioms(self, all_classes, type_sort):
+        """
+        Creates axioms defining subtype relations for all possible classes.
+        """
+        tree = self.create_class_tree(all_classes, type_sort)
+        axioms = []
+        # For each class C in the program, create two axioms:
+        for c in tree.all_children():
+            c_literal = c.get_literal()
+            x = Const("x", self.type_sort)
+
+            # One which is triggered by subtype(C, X)
+            options = []
+            for sub in c.all_super():
+                options.append(x == sub.get_literal())
+            subtype_expr = self.subtype(c_literal, x)
+            axiom = ForAll([x] + c.quantified(), subtype_expr == Or(*options),
+                           patterns=[subtype_expr])
+            axioms.append(axiom)
+
+            # And one which is triggered by subtype(X, C)
+            options = []
+            for super in c.all_children():
```

Comment:
This gives an annoying warning about shadowing the built-in name `super`. Can you rename it? 😇

Also, should not this be `sub`, instead?